AztecProtocol/aztec-v1
# Added Wallet contract

## Description


Added a Wallet contract. This contract owns a user's notes on behalf of them, and notes can be approved, transferred, and these transactions signed. The idea of this is that, as contracts have no private or public key, they can sign and approve batches of notes together. This means a user does not have to individually sign every note in a transaction where multiple notes are used in a transaction (e.g. in a join-split).

## Testing instructions

`yarn test` on packages/protocol/

Key methods are:
- the constructor
- `batchConfidentialApprove(bytes32[] memory _noteHashes, address _zkAsset, address _spender, bool _status)`
- `batchConfidentialTransfer(bytes memory _proof, address _zkAsset, address _sender)`
- `batchValidateSignature(bytes32[] memory _noteHashes, address _spender, bool _status, bytes memory _signature, address _zkAsset)`

Potential adjustments before production depending on how extension team want flow of wallet to work:
- signing function may want to be brought inside the approval function
- **some functions may want to be called inside others and made internal**, so there is just one big spend functions - the `spendNotes` method is a prototype attempt at this, **it should be decided whether the `spendNotes` function should remain before production**
- extra tests may be required (e.g. testing empty arrays of noteHashes though this is an unlikely situation)

Changes possibly affecting other parts of protocol:
- Testing and coverage now run ganache-cli with the same menmonic every time, set in the envrionment variable $TEST_MNEMONIC. This is neccesary because, to test signing, the signer must own the wallet contract, so we must be able to get the private key of an actual truffle user. These keys/addresses are derived from the $TEST_MNEMONIC environment variable. The other (non-signing) tests could be adjusted to work without this by changing the zkAsset and wallet contract owner, but in their current state also require this known mnemonic.
- scripts/test.sh starts ganache with this mnemonic, either using a preset environment variable, or from a .env file, or from a .env.example file. If none of these sources are present tests will fail.

## Types of changes





* New feature (non-breaking change which adds functionality)



## Checklist:
